viz/cmd/stat-inbound.go

```diff
@@ -0,0 +1,411 @@
+
+			results := make(map[inboundRowKey]inboundRow)
+
+			for sample := range responseChan {
+				labels := sample.Metric
+				key := inboundKeyForSample(sample, resource)
+				row := results[key]
+				if labels["classification"] == "success" {
+					row.successes += uint64(sample.Value)
+				} else {
+					row.failures += uint64(sample.Value)
+				}
+				results[key] = row
+			}
+
+			for quantile, resultsChan := range quantiles {
+				for sample := range resultsChan {
+					key := inboundKeyForSample(sample, resource)
+					row := results[key]
+					row.populateLatency(quantile, sample)
+					results[key] = row
+				}
+			}
+
+			// Render output.
+			timeWindow, err := time.ParseDuration(options.timeWindow)
+			if err != nil {
+				return err
+			}
+			if options.outputFormat == "json" {
+				return renderStatInboundJson(results, timeWindow)
+			} else if options.outputFormat == "table" || options.outputFormat == "" {
+				return renderStatInboundTable(results, timeWindow)
+			} else {
+				return fmt.Errorf("Invalid output format: %s", options.outputFormat)
+			}
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.outputFormat, "output", "o", options.outputFormat, "Output format; one of: \"table\" or \"json\" or \"wide\"")
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace")
+	cmd.PersistentFlags().StringVarP(&options.timeWindow, "time-window", "t", options.timeWindow, "Stat window (for example: \"10s\", \"1m\", \"10m\", \"1h\")")
+	cmd.PersistentFlags().StringVar(&options.prometheusURL, "prometheusURL", options.prometheusURL, "Address of Prometheus instance to query")
+
+	return cmd
+}
+
+func queryRate(
+	ctx context.Context,
+	promAPI promv1.API,
+	metric string,
+	timeWindow string,
+	labels model.LabelSet,
+	groupBy model.LabelNames,
+	resource *pb.Resource,
+) <-chan *model.Sample {
+	results := make(chan *model.Sample)
+	go func() {
+		defer close(results)
+		query := fmt.Sprintf("sum(increase(%s%s[%s])) by (%s)",
+			metric,
+			labels.Merge(prometheus.QueryLabels(resource)),
+			timeWindow,
+			append(groupBy, prometheus.GroupByLabelNames(resource)...),
+		)
+		log.Debug(query)
+		val, warn, err := promAPI.Query(ctx, query, time.Time{})
+		if warn != nil {
+			log.Warnf("%v", warn)
+		}
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "failed to query Prometheus: ", err.Error())
+			os.Exit(1)
+		}
+		for _, sample := range val.(model.Vector) {
+			results <- sample
+		}
+	}()
+	return results
+}
+
+func queryQuantiles(
+	ctx context.Context,
+	promAPI promv1.API,
+	metric string,
+	timeWindow string,
+	labels model.LabelSet,
+	groupBy model.LabelNames,
+	resource *pb.Resource,
+) map[string]chan *model.Sample {
+	results := map[string]chan *model.Sample{
+		"0.5":  make(chan *model.Sample),
+		"0.95": make(chan *model.Sample),
+		"0.99": make(chan *model.Sample),
+	}
+	for quantile, resultsChan := range results {
+		go func(quantile string) {
+			defer close(resultsChan)
```

Comment:
Apparently not since all the quantile channels do get populated with this code.  I think because channels are always passed by reference, the contents of the loop variable are already a reference (instead of being a reference to the loop variable).

Reply:
hmm I think that being a reference doesn't matter for how closures use those vars, _but_ we likely are just benefiting from [go 1.22's fix around this type of issue](https://tip.golang.org/wiki/LoopvarExperiment), so nothing to see here :-)